error_handler.py

```diff
@@ -0,0 +1,114 @@
+import datetime
+import io
+import logging
+import traceback
+import urllib
+
+import discord
+from discord.ext import commands
+
+logger = logging.getLogger(__name__)
+
+
+def get_full_class_name(obj):
+    module = obj.__class__.__module__
+    if module is None or module == str.__class__.__module__:
+        return obj.__class__.__name__
+    return module + '.' + obj.__class__.__name__
+
+class NonBugError(Exception):
+    """When this is raised instead of a normal Exception, on_command_error() will not attach a traceback or github
+    link. """
+    pass
+
+# completely overkill error handler shamelessly ripped from MediaForge, a bot i wrote
+class ErrorHandlerCog(commands.Cog):
+    def __init__(self, bot):
+        self.bot: commands.Bot = bot
+        self.antispambucket = {}
+
+    @commands.Cog.listener()
+    async def on_command_error(self, ctx: commands.Context, commanderror: commands.CommandError):
+        async def dmauthor(*args, **kwargs):
+            try:
+                return await ctx.reply(*args, **kwargs)
+            except discord.Forbidden:
+                logger.info(f"Reply to {ctx.message.id} and dm to {ctx.author.id} failed. Aborting.")
+
+        async def reply(msg, file=None, embed=None):
+            if ctx.interaction:
+                if ctx.interaction.response.is_done():
+                    return await ctx.interaction.edit_original_response(content=msg,
+                                                                        attachments=[file] if file else None,
+                                                                        embed=embed)
+                else:
+                    return await ctx.reply(msg, file=file, embed=embed)
+            else:
+                try:
+                    if ctx.guild and not ctx.channel.permissions_for(ctx.me).send_messages:
```

Comment:
this is a TOCTOU bug, since permissions can change between us checking and sending the reply, we should attempt to send the message always and catch errors it gives